### PR TITLE
fix(metrics): Re-initialize global tags in subprocesses of the parallel indexer

### PR DESCRIPTION
**Pull request by untitaker**
This fixes tagging in subprocesses, and we already have a smoke test
that the indexer doesn't crash in principle (though I verified locally
that the indexer comes up fine on an empty topic)

### CODE REVIEW

File: src/sentry/sentry_metrics/configuration.py [hunk 1]
```diff
@@ -131,3 +136,41 @@ def get_ingest_config(
...
+    """
+
+    import sentry_sdk
+
+    sentry_sdk.set_tag("sentry_metrics.use_case_key", config.use_case_id.value)
+
+    from sentry.utils.metrics import global_tags
+
+    ctx = global_tags(_all_threads=True, pipeline=config.internal_metrics_tag)
+    ctx.__enter__()
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/configuration.py:176]: nobody is ever exiting this context ?

**Review by fpacifici — COMMENTED** [review]
I think there is another problem here 
https://github.com/getsentry/sentry/blob/master/src/sentry/runner/initializer.py#L405

The arroyo metrics would not have the pipeline tag either. 

(also that initialization that is valid across sentry would add `indexer` into the metric name for whoever uses arroyo, which seems wrong)

File: src/sentry/sentry_metrics/configuration.py (hunk 1 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/configuration.py:176]: We could register an atexit handler, but I think that's fine to be honest. We will preserve the global tags for the lifetime of the process.

**Comment by untitaker** [comment]
ooh, that makes a lot of sense. let me know what you think of this version then

File: src/sentry/sentry_metrics/configuration.py [hunk 2]
```diff
@@ -131,3 +136,51 @@ def get_ingest_config(
...
+def initialize_global_consumer_state(config: MetricsIngestConfiguration) -> None:
+    """
+    Initialization function for the main process of the metrics indexer.
+
+    This primarily sets global tags for instrumentation in both our
+    statsd/metrics usage and the Sentry SDK.
+    """
+
+    import sentry_sdk
+    from arroyo import configure_metrics
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/configuration.py:170]: What is the reason for not importing this at the top of the module ? It should not be a problem for the multi process consumer.

**Review by fpacifici — APPROVED** [review]
Please avoid using the context manager without exiting. 
The context manager interface is pretty clear (for whoever writes the context manager) you enter and then you exit, so who writes that context manager can rely on this property. If we abuse it we look for troubles even if in its current implementation, in our case this does not end up causing problems right now. 

Since what you want to do is to initialize the tags and you are fine leaving them in place till the end, please have a function in `utils/metrics.py` that does the same thing without the expectations that it would be cleaned up at the end. This at least sets up a contract we are honoring.

File: src/sentry/sentry_metrics/configuration.py (hunk 1 above)
Comment (fpacifici) [reply src/sentry/sentry_metrics/configuration.py]: See my comment below.

File: src/sentry/sentry_metrics/configuration.py (hunk 2 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/configuration.py:170]: ok, i moved those two imports to the t

**Comment by untitaker** [comment]
> so who writes that context manager can rely on this property

I don't think this is generally true. you can use a context manager in a generator function. if the generator is not exhausted, the context manager is leaked this way.

still, point taken about the interface, i made some changes to `sentry.utils.metrics` that hopefully make it easier to use.

**Pull request merged**